### PR TITLE
Mint Pixel Avatars layer 2

### PR DESCRIPTION
**Pull request by rasmuscnielsen**
This PR is a working example of allowing the Pixel avatar contract to be hosted on ie Polygon. It uses a trusted webserver to authorize a claim with a signature that is sent to the contract.

The workflow is: 

- User signs in with their address in web client
- Client lists available token ids from genesis contract (in fact this functionality was moved to the server too and is exposed with API call)
- User selects token to mint. 
- Client sends authorize request to server containing address + token id
- Server checks the address holds genesis token. It creates a signature with: address, tokenId, deadline (now + 5min to mint)
- Client forwards the server signature along with raw parameters
- The solidity contract verifies that the signature originates from the trusted server (checking address), and verifies the minting address + token matches the signed message

Would love feedback in the general approach as well as the contract implementation as this is my very first time writing Solidity.

### CODE REVIEW

**Comment by Dhaiwat10** [comment]
These things might or might not be relevant to this very PR but I hope it's some useful information:

- The site didn't tell me to change to the correct network when I was on mainnet. This is important.
- In case someone doesn't own any genesis NFTs, an appropriate message should be shown. Eg. 'Could not detect a genesis NFT that you own.'
image

File: pixel-avatar-hardhat/contracts/PixelAvatars.sol [hunk 1]
```diff
@@ -64,60 +30,59 @@ contract PixelAvatars is ERC721Enumerable, ReentrancyGuard, Ownable {
...
-        payable(msg.sender).transfer(address(this).balance);
+        _;
     }
 
-    function multiMintWithDevDaoToken(uint256[] memory _devDaoTokenIds)
-        public
-        payable
-        nonReentrant
-        multipleDevDaoTokenOwnerOf(_devDaoTokenIds)
+    function mintWithSignature(
```
Comment (gjsyme) [inline pixel-avatar-hardhat/contracts/PixelAvatars.sol:66]: Function looks good, but the business logic was previously supporting minting more than one token per transaction. Looks like we can now only mint a single token per transaction and I'm not sure if that is a desired change.

**Review by gjsyme — APPROVED** [review]
Looks good to me. Biggest thing that stands out is that we went from multi-mint to only supporting single mint, but that's a business decision not code. Another nitpick is the use of Promise.all; I didn't see anything that would handle a failure, but I'm also not certain there's a case where you would get one failure with others returning as expected so it may be a moot point.

**Comment by rasmuscnielsen** [comment]
@Dhaiwat10 Thanks for the feedback! 

I believe these issues will be resolved when deployed. Right now the deployment isn't working properly as the contract is yet to be deployed so it can be properly tested. 

Once done, I'll make sure to see if the issues are resolved!

**Comment by rasmuscnielsen** [comment]
@gjsyme 

Thanks for the feedback and code review - much appreciated! 

> Biggest thing that stands out is that we went from multi-mint to only supporting single mint, but that's a business decision not code

Completely agree, and definitely something that will need consensus from the working group. 

From a business perspective I do believe it is the better option not to offer multi-mint, as this would basically allow people to mint 100x avatars for the price of one. If anything, it should make sure that the user is charged accordingly for the amount of avatars they are minting. 
Overall though I believe the DAO isn't too concerned to cater to whales that snatched up bulks of genesis tokens. I even remember few suggestions around limiting how many avatars an address can hold to 1 regardless of the amount of genesis tokens held. 

But really valuable comment, and 👍 for initiating the conversation.

**Comment by rasmuscnielsen** [comment]
@gjsyme @Dhaiwat10 

Just did a bunch of work on this tonight, and finally managed to get everything deployed on Mumbai testnet + Vercel 🎉 

Since I don't have access to the DAO Vercel, I got it up and running here instead: https://pixel-avatars-jet.vercel.app/ 

@Dhaiwat10 regarding

> The site didn't tell me to change to the correct network when I was on mainnet. This is important.

I tested this a bunch, and it will display an error messages on connect + on minting attempts when on wrong Network. 
Furthermore I managed to do a window refresh, BUT this only works on MetaMask in the browser. I can't seem to get this working for Wallet Connect. Some more testing might be needed what happens if a user changes network in wallet when already connect via Wallet Connect. It'd be great if someone could dig more into this, and perhaps provide a fix.

@gjsyme regarding

> Biggest thing that stands out is that we went from multi-mint to only supporting single mint, but that's a business decision not code

I had a good chat with @goldzulu about this (meeting notes should be on their way), and we both agreed multi-mint isn't something we wish to go out of our way to support. 
Obviously anyone and everyone are free to challenge this standpoint, but until someone speaks up we'll continue on without it.

**TODO FROM HERE**

- Have this solution peer-reviewed by someone with a good understanding of Solidity and signatures
- Get the DAO Vercel up and running. It might be a little bit tricky to get the "previews" working given it's a mono-repo and the "previewer client" is dependent on the server. Also, we might not be able to get the server hosted before it's merged to Main. 
- Coordinate release with DAO 
- Deploy contract to production (incl. generate fresh private/public key for server)
-  **Important:** We have currently settled on a miniting price that is 0.01eth. However if we deploy to MATIC, we should change this accordingly - otherwise our avatars will be VERY cheap :-)

**Comment by gjsyme** [comment]
> I can't seem to get this working for Wallet Connect

So I don't have all the setup done to run this code locally, but on your PR I think that you should be able to see the network of a @walletconnect/web3-provider after completing the `await this.provider.enable()` call on line 17 as `this.provider.provider.chainId` (at least, that's what's showing for me on a mint site I have). At the very least you can take action / disconnect / reload the whole page as required at that point (if you haven't already considered this).

**Comment by Dhaiwat10** [comment]
@rasmuscnielsen one more suggestion:

image

Instead of giving the user a link to the Opensea profile, we should instead give them a link directly to the asset. The Opensea format is: `https://testnets.opensea.io/assets/mumbai/{contractAddress}/{tokenId}`

**Comment by Dhaiwat10** [comment]
I would also add a message telling the user which network they need to connect to and a button that lets them add Polygon network to their Metamask.

Apart from that, it worked pretty well. Awesome job.

**Comment by briangershon** [comment]
Hey @rasmuscnielsen

- [ ] I noticed that as you cycle through developer IDs in the UI, they no longer update query string in the URL. This was handy so that users can grab the URL for a specific NFT. e.g. The production version shows a URL like `https://pixel-avatars.developerdao.com/?developerId=6`.

- [ ] I noticed because I fixed a bug where visiting a URL wasn't displaying the write NFT (#52). See PR #53 for the one-line fix for that.

**Comment by rasmuscnielsen** [comment]
@gjsyme @Dhaiwat10 

**Regarding network:** Thanks for your feedback! I made a bunch of improvements, and the network error is now much more clear. There are still room for improvement (suggestions + PR's welcome), but I believe what's currently live on https://pixel-avatars-jet.vercel.app/ is pretty good starting point for now.  

Skærmbillede 2021-12-19 kl  21 23 54

____

**Regarding linking directly to OpenSea assets:** I tried this out locally, and the issue seems to be that OpenSea relies pretty heavily on caching. For the first 12-24 hours or so, it just displayed a 404 🙁 I tried in several different browsers as well with the same result.
Skærmbillede 2021-12-19 kl  23 35 43

____

@briangershon Thanks for your PR and raising this issue. It should now be fixed on https://pixel-avatars-jet.vercel.app/ !
 
 ____
 
 **PROPOSAL FOR NEXT STEPS**
 
I suggest to merge this PR to `main` branch as is. I have made a if-statement that hides the Connect-dropdown if the chain-id is not set, so this will allow us to put the current visual improvements live, and prevent this PR for lagging further behind `main` branch. 
 
I then suggest that we make issues for further improvements. This will also help people that want to contribute to get in on the action. 

Once it's merged, I also propose to make a new `staging` branch where we can setup the `env` vars to use the Mumbai testnet for further testing. 
For this I'll be needing your help @Dhaiwat10 to setup Vercel properly, as we're using a mono-repo approach and will need to configure an additional 'webserver' project pointing to this repo :-) 

Sounds good?

**Comment by briangershon** [comment]
Thanks @rasmuscnielsen for your work on this, and fixes for the query string items!

I was in the process of getting this project setup locally earlier this week, but had to take a Solidity development detour to ramp up.

I won't be able to attend the Derivatives meeting today (will be traveling on the road), but will be setting this up locally and learning more about it as I go...

**Comment by Dhaiwat10** [comment]
> For this I'll be needing your help @Dhaiwat10 to setup Vercel properly, as we're using a mono-repo approach and will need to configure an additional 'webserver' project pointing to this repo :-) Sounds good?

Yeah sounds good to me! Tag me whenever you need help with Vercel

**Comment by goldzulu** [comment]
@rasmuscnielsen feel free to reorganise the directory structure to make more sense if needed

**Pull request merged**